graph/graph.go

```diff
@@ -535,12 +631,48 @@ func jsonPath(root string) string {
 	return filepath.Join(root, jsonFileName)
 }
 
+// storeImage stores file system layer data for the given image to the
+// graph's storage driver. Image metadata is stored in a file
+// at the specified root directory.
+func (graph *Graph) storeImage(img *image.Image, layerData archive.ArchiveReader, root string) (err error) {
+	// Store the layer. If layerData is not nil, unpack it into the new layer
+	if layerData != nil {
+		if err := graph.disassembleAndApplyTarLayer(img, layerData, root); err != nil {
+			return err
+		}
+	}
+
+	if err := graph.saveSize(root, int(img.Size)); err != nil {
+		return err
+	}
+
+	f, err := os.OpenFile(jsonPath(root), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
+	if err != nil {
+		return err
+	}
+
+	defer f.Close()
+
+	return json.NewEncoder(f).Encode(img)
+}
+
 func (graph *Graph) disassembleAndApplyTarLayer(img *image.Image, layerData archive.ArchiveReader, root string) error {
 	// this is saving the tar-split metadata
 	mf, err := os.OpenFile(filepath.Join(root, tarDataFileName), os.O_CREATE|os.O_WRONLY|os.O_TRUNC, os.FileMode(0600))
 	if err != nil {
 		return err
 	}
+
+	// TODO WINDOWS: The metadata cannot be used on save of a layer,
+	// so remove it in order to fallback to normal Diff behavior.
+	if runtime.GOOS == "windows" {
```

Comment:
We can... but it might make behavior of the method different enough on the two platforms that the method implementation should be split, which would land us right back at having graph_windows and graph_unix. Let me see if I can make a reasonable enough version that avoids the file creation without needing to split the implementation, and we can keep this deferred deletion as a simple fallback if that doesn't work out.